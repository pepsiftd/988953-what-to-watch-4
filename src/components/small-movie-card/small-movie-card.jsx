import React from 'react';
import PropTypes from 'prop-types';
import {history} from '@/history';
import {AppRoute} from '@/const';
import {Link} from 'react-router-dom';

const PREVIEW_START_TIMEOUT = 1000;

const SmallMovieCard = ({movie, onMouseEnter, onMouseLeave, children}) => {
  const {title, id} = movie;
  let timeout = null;

  return (
    <article className="small-movie-card catalog__movies-card"
      onMouseEnter={() => {
        timeout = setTimeout(() => {
          onMouseEnter(id);
        }, PREVIEW_START_TIMEOUT);
      }}
      onMouseLeave={() => {
        if (timeout) {
          clearTimeout(timeout);
        }
        onMouseLeave();
      }}
      onClick={() => {
        history.push(`${AppRoute.MOVIE_PAGE}/${id}`);
      }}
    >
      <div className="small-movie-card__image">
        {children}
      </div>
      <h3 className="small-movie-card__title">
        <Link className="small-movie-card__link" to={`${AppRoute.MOVIE_PAGE}/${id}`}>{title}</Link>
      </h3>
    </article>
  );
};

SmallMovieCard.propTypes = {
  movie: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    imageSrc: PropTypes.string.isRequired,
    preview: PropTypes.string.isRequired,
  }).isRequired,
  children: PropTypes.element.isRequired,
  onMouseEnter: PropTypes.func.isRequired,
  onMouseLeave: PropTypes.func.isRequired,
};

export {SmallMovieCard};
