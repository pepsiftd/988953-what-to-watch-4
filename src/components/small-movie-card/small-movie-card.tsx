import React from 'react';
import {history} from '@/history';
import {AppRoute} from '@/const';
import {Link} from 'react-router-dom';

const PREVIEW_START_TIMEOUT = 1000;

interface Props {
  movie: {
    id: number;
    title: string;
    imageSrc: string;
    preview: string;
  };
  children: React.ReactElement;
  onMouseEnter: (id: number) => void;
  onMouseLeave: () => void;
}

const SmallMovieCard: React.FunctionComponent<Props> = ({movie, onMouseEnter, onMouseLeave, children}) => {
  const {title, id} = movie;
  let timeout = null;

  return (
    <article className="small-movie-card catalog__movies-card"
      onMouseEnter={() => {
        timeout = setTimeout(() => {
          onMouseEnter(id);
        }, PREVIEW_START_TIMEOUT);
      }}
      onMouseLeave={() => {
        if (timeout) {
          clearTimeout(timeout);
        }
        onMouseLeave();
      }}
      onClick={() => {
        history.push(`${AppRoute.MOVIE_PAGE}/${id}`);
      }}
    >
      <div className="small-movie-card__image">
        {children}
      </div>
      <h3 className="small-movie-card__title">
        <Link className="small-movie-card__link" to={`${AppRoute.MOVIE_PAGE}/${id}`}>{title}</Link>
      </h3>
    </article>
  );
};

export {SmallMovieCard};
