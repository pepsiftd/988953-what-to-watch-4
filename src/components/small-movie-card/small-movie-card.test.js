import React from 'react';
import renderer from 'react-test-renderer';
import {SmallMovieCard} from './small-movie-card';
import {Router} from 'react-router-dom';
import {history} from '@/history';

const movie = {
  id: 152,
  title: `Some movie: Revenge`,
  imageSrc: `img/some-movie-revenge.jpg`,
  movieLink: `movie-page.html`,
  preview: `preview-link.mp4`,
};

it(`SmallMovieCard should render correct movie name`, () => {
  const tree = renderer.create(
      <Router history={history}>
        <SmallMovieCard
          movie={movie}
          onMouseEnter={() => {}}
          onMouseLeave={() => {}}
        >
          <video />
        </SmallMovieCard>
      </Router>
  );

  expect(tree).toMatchSnapshot();
});
