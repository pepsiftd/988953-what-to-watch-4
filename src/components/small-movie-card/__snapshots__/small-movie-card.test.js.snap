// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmallMovieCard should render correct movie name 1`] = `
<article
  className="small-movie-card catalog__movies-card"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="small-movie-card__image"
  >
    <video />
  </div>
  <h3
    className="small-movie-card__title"
  >
    <a
      className="small-movie-card__link"
      href="/films/152"
      onClick={[Function]}
    >
      Some movie: Revenge
    </a>
  </h3>
</article>
`;
